src/Util/TestResultCache.php

```diff
@@ -0,0 +1,127 @@
+<?php
+/*
+ * This file is part of PHPUnit.
+ *
+ * (c) Sebastian Bergmann <sebastian@phpunit.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace PHPUnit\Runner;
+
+use PHPUnit\Framework\ResultCacheListener;
+
+class TestResultCache
+{
+    /**
+     * @var string
+     */
+    public const DEFAULT_RESULT_CACHE_FILENAME = 'result_cache.json';
+
+    /**
+     * @var array
+     */
+    private $defects = [];
+
+    /**
+     * @var array
+     */
+    private $times = [];
+
+    public function getResultCacheFilename()
+    {
+        return $_ENV['PHPUNIT_RESULT_CACHE'] ?? self::DEFAULT_RESULT_CACHE_FILENAME;
+    }
+
+    public function persist(): void
+    {
+        if (isset($_ENV['PHPUNIT_RESULT_CACHE_SELFTEST'])) {
+            \var_dump($this->formatJSONOutputStruct());
+
+            return;
+        }
+
+        $this->saveToFile();
+    }
+
+    public function saveToFile(): void
+    {
+        \file_put_contents(
+            $this->getResultCacheFilename(),
+            $this->formatJSONOutputStruct()
+        );
+    }
+
+    public function getState($testName): string
+    {
+        return $this->defects[$testName] ?? ResultCacheListener::TEST_UNKNOWN;
+    }
+
+    public function getTime($testName): float
+    {
+        return $this->times[$testName] ?? 0;
+    }
+
+    public function loadFromListenerRunstate(array $runState): void
+    {
+        $this->clearCache();
+
+        // Only store tests with non-PASS results
+        $this->defects = \array_map(
+            function ($result) {
+                return $result['state'];
+            },
+            \array_filter($runState, function ($result) {
+                return !\in_array($result['state'], ['suite', 'passed'], true);
+            })
+        );
+
+        $this->times = \array_map(function ($result) {
+            return $result['time'];
+        }, $runState);
+    }
+
+    public function loadFromFile(): void
+    {
+        $this->clearCache();
+
+        if (\is_file($this->getResultCacheFilename()) === false) {
+            return;
+        }
+
+        $json = \file_get_contents($this->getResultCacheFilename());
+
+        if ($json === false) {
+            return;
+        }
+
+        $this->loadFromJSON($json);
+    }
+
+    public function loadFromJSON(string $json)
+    {
+        $cacheData = \json_decode($json, true);
+
+        if ($cacheData === null) {
+            return;
+        }
+
+        $this->defects = $cacheData['defects'];
```

Comment:
we have no clue that decoded data contain such key...
and even if - what is the content of it.

it's optimistic approach here.

cache file could be treated as-is, one is not supposed to modify it manually.
maybe serialize with `unserialize(..., ['allowed_classes' => CacheData::class])` ?

Reply:
Yes, the earliest versions of Breakfast were just that: patches to the code that quickly (un)serialized a results file to store the state between runs. I'll try some other options.